### DISCUSSION TITLE
Self Hosted Runners

### DISCUSSION THREAD

**Discussion by nastanford** [community/community#27336] [Actions]
## How can you make a self-hosted runner be called in the YML file by the name?

**Example:**
I am wanting to make a button where when someone commits it will push to staging and then somehow push to production later.
I run the runner on the server I want it pushed to so there are no security concerns. 
I am new to GitHub actions so my process may not be correct just yet.  I am not using SSH to connect to the machine as we don't want any outside service to SSH into it. 

Thanks for any suggestions or help.

**Comment by airtower-luna** [comment]
You can't, only by label. Of course you could technically use a label that applies to that machine only, but "deploy by runner" is not a very good solution in most cases. It's usually better to use an automation tool (e.g. Ansible, Puppet, ...) and run that from the workflow.

> I am not using SSH to connect to the machine as we don't want any outside service to SSH into it.

With a self-hosted runner on the target network (but not necessarily the same machine) you wouldn't need any SSH connection from the outside. And I know from experience that Ansible works beautifully over SSH.

**Reply by nastanford** [reply]
Thanks, for the advice What is the "Normal" way to do this?  Where I am we do not always do what may seem normal but I like to get as close as possible.  What of Jenkins? what other tools are the best to research?

**Reply by airtower-luna** [reply]
My main point is that you shouldn't use the Actions runner directly to make *persistent* changes to the host it's running on. Runner hosts are meant to be replaceable. Calling an automation tool in the Actions workflow to configure a specific host decouples your deployment from the runner instance.

I am extremely happy with Ansible and use it personally and for work, the playbooks are easy to write (YAML syntax), and it just needs SSH to work. But what's the best tool depends on your environment (e.g. Ansible is great with Linux, but somewhat limited with Windows, AFAIK). Of course if your deployment is simple you could also roll your own, e.g. SSH into the target system, unpack some artifact, atomically move it into place.

Jenkins is a CI system, so an alternative to Actions. I'd advise against using a Jenkins node to deploy to the node itself just as much as I advise against doing it with the Actions runner. :wink:

**Comment by daltonv** [comment]
So @airtower-luna is right that you can only call runners by labels on not by the runner name itself, but I think this should be a feature that is added. 

I currently have some self-hosted windows runners that I manually manage. Sometimes I want to test that a workflow works on one of the machines after I made some changes. I currently handle this by making labels that are the same name of the runner, but I think this could be a bit cleaner if we could get a built in capability to call the self-hosted runners by their name.
